### PR TITLE
@orta: Transition largest VC to MVVM

### LINKED ISSUE #333: MVVMification
We've got a lot of heavy view controllers (looking at you, ListingsViewController). We should move the code into view models where possible, writing tests as we go. (Since our VCs are woefully undertested right now, this is a great excuse to write some!)

This should cover work from https://github.com/artsy/eidolon/issues/176 too.

### PR DESCRIPTION
**Pull request by ashfurrow**
This is currently a straight "move code into a new file, refer to it there" change. Ideally, the VC won't even have access to the sale artworks arrays. 

Fixes #333, ish.

### CODE REVIEW

**Comment by ashfurrow** [comment]
So, things are in an OK state now. Things I still need to do:

- [ ] More clearly delineate responsibilities of the vm/vc – for example, I don't like that the view controller can access the sale artworks directly. Could use a protocol for easy stubbing, too.
- [ ] Look through view model for opportunities for better code, style, and comments.
- [ ] Get existing unit tests working, duh.
- [ ] Write tests for new view model.

**Comment by ashfurrow** [comment]
So this is in a _pretty_ good place. Remaining issues:

- [ ] Some of the LoadingViewModelTests are intermittently failing, not sure why. When run as focused tests (`fit`), they always pass. However, they're timing out otherwise. Could be an Xcode 7 thing, no open issues on Quick.
- [ ] Writing tests for new view model (some of the old listings VC test code is there now, commented out.

File: Kiosk/App/Models/SaleArtworkViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        default:
+            return "No Estimate"
+        }
+    }
+
+    var thumbnailURL: NSURL? {
+        return saleArtwork.artwork.defaultImage?.thumbnailURL()
+    }
+
+    var thumbnailAspectRatop: CGFloat? {
```
Comment (orta) [inline Kiosk/App/Models/SaleArtworkViewModel.swift]: `thumbnailAspectRatop`

File: Kiosk/Admin/ChooseAuctionViewController.swift [hunk 2]
```diff
@@ -25,7 +25,7 @@ class ChooseAuctionViewController: UIViewController {
 
                     let button = ARFlatButton()
                     button.setTitle(title, forState: .Normal)
-                    button.setTitleColor(UIColor.blackColor(), forState: .Normal)
```
Comment (orta) [inline Kiosk/Admin/ChooseAuctionViewController.swift]: I like this

File: Kiosk/App/GlobalFunctions.swift [hunk 3]
```diff
@@ -31,3 +31,28 @@ func responseIsOK(object: AnyObject!) -> AnyObject {
...
+    if if0 == nil { return nil }
+
+    let interfaceName: CFStringRef = unsafeBitCast(if0!, CFStringRef.self)
+    let dictionary = CNCopyCurrentNetworkInfo(interfaceName) as NSDictionary?
+    if dictionary == nil { return nil }
+
+    return dictionary?[kCNNetworkInfoKeySSID as String] as? String
+}
+
+func detectDevelopment() -> Bool {
```
Comment (orta) [inline Kiosk/App/GlobalFunctions.swift]: I'm not sure that this function name is entirely descriptive, `detectDevelopmentEnvironment` ?

File: Kiosk/Auction Listings/ListingsViewController.swift [hunk 4]
```diff
@@ -371,86 +204,21 @@ private extension ListingsViewController {
...
-        case LowestCurrentBid:
-            return saleArtworks.sort(lowestCurrentBidSort)
-        case Alphabetical:
-            return saleArtworks.sort(alphabeticalSort)
-        }
-    }
-    
-    static func allSwitchValues() -> [SwitchValues] {
-        return [Grid, LeastBids, MostBids, HighestCurrentBid, LowestCurrentBid, Alphabetical]
+    class func listingsCollectionViewWithDelegateDatasource<T where T: UICollectionViewDelegate, T: UICollectionViewDataSource>(delegateDatasource: T) -> UICollectionView {
```
Comment (orta) [inline Kiosk/Auction Listings/ListingsViewController.swift]: I don't understand the need for all the generics here?

File: Kiosk/Auction Listings/ListingsViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        self.showDetails = showDetails
+        self.presentModal = presentModal
+
+        super.init()
+
+        RAC(self, "saleArtworks") <~ recurringListingsRequestSignal()
+
+        showSpinnerSignal = RACObserve(self, "saleArtworks").mapArrayLengthExistenceToBool().not()
+        gridSelectedSignal = selectedIndexSignal.map { (index) -> AnyObject! in
+            switch ListingsViewModel.SwitchValues(rawValue: index as! Int) {
```
Comment (orta) [inline Kiosk/Auction Listings/ListingsViewModel.swift]: could 68-72 turn into a ternary?

File: Kiosk/Auction Listings/ListingsViewController.swift (hunk 4 above)
Comment (ashfurrow) [inline Kiosk/Auction Listings/ListingsViewController.swift]: Couldn't find a way to have a parameter that must conform to two protocols without making it a generic with constraints.

File: Kiosk/Auction Listings/ListingsViewModel.swift (hunk 5 above)
Comment (ashfurrow) [inline Kiosk/Auction Listings/ListingsViewModel.swift]: Good idea.

File: Kiosk/Bid Fulfillment/YourBiddingDetailsViewController.swift [hunk 6]
```diff
@@ -21,7 +20,7 @@ class YourBiddingDetailsViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        $.each([notificationLabel, bidderNumberLabel, pinNumberLabel]) { $0.makeTransparent() }
+        [notificationLabel, bidderNumberLabel, pinNumberLabel].forEach { $0.makeTransparent() }
```
Comment (orta) [inline Kiosk/Bid Fulfillment/YourBiddingDetailsViewController.swift]: might even be able to remove `$` from the codebase?

Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/YourBiddingDetailsViewController.swift]: Done :tada:

File: Kiosk/Auction Listings/ListingsViewController.swift (hunk 4 above)
Comment (a2) [inline Kiosk/Auction Listings/ListingsViewController.swift]: `protocol< UICollectionViewDelegate, UICollectionViewDataSource>`

Comment (ashfurrow) [inline Kiosk/Auction Listings/ListingsViewController.swift]: I was hoping it wouldn't come to that :crying_cat_face:

Comment (tapi) [inline Kiosk/Auction Listings/ListingsViewController.swift]: Ran into this at WWDC2014, the guys there told me the same thing as @a2

Comment (ashfurrow) [inline Kiosk/Auction Listings/ListingsViewController.swift]: Alright I'll do that. Thanks!

**Comment by ashfurrow** [comment]
screen shot 2015-09-03 at 6 17 23 pm

@orta ok this is good to review. The final three commits should catch you up.

**Comment by orta** [comment]
Great, looks good

**Pull request merged**